File: UI/clinicnow_desktop/lib/models/lab_finding.dart
/// Mirrors the backend's `LabFindingDto`. Never carries raw file bytes - only
/// [downloadUrl] (the actual bytes are served from a dedicated endpoint).
class LabFinding {
  final int id;
  final int patientId;
  final String patientName;
  final int appointmentId;
  final DateTime appointmentStartUtc;
  final String medicalServiceName;
  final String testName;
  final String? value;
  final String? unit;
  final String? referenceRange;
  final String result;
  final String? doctorNote;
  final String? fileName;
  final String? contentType;
  final int fileSizeBytes;
  final String enteredByName;
  final DateTime createdAtUtc;
  final String downloadUrl;

  /// Whether a document is attached at all - the attachment is optional, so
  /// the "Preuzmi" action is only offered when this is true.
  final bool hasFile;

  const LabFinding({
    required this.id,
    required this.patientId,
    required this.patientName,
    required this.appointmentId,
    required this.appointmentStartUtc,
    required this.medicalServiceName,
    required this.testName,
    this.value,
    this.unit,
    this.referenceRange,
    required this.result,
    this.doctorNote,
    this.fileName,
    this.contentType,
    required this.fileSizeBytes,
    required this.enteredByName,
    required this.createdAtUtc,
    required this.downloadUrl,
    required this.hasFile,
  });

  factory LabFinding.fromJson(Map<String, dynamic> json) {
    return LabFinding(
      id: json['id'] as int,
      patientId: json['patientId'] as int,
      patientName: json['patientName'] as String? ?? '',
      appointmentId: json['appointmentId'] as int,
      appointmentStartUtc: DateTime.parse(json['appointmentStartUtc'] as String),
      medicalServiceName: json['medicalServiceName'] as String? ?? '',
      testName: json['testName'] as String? ?? '',
      value: json['value'] as String?,
      unit: json['unit'] as String?,
      referenceRange: json['referenceRange'] as String?,
      result: json['result'] as String? ?? '',
      doctorNote: json['doctorNote'] as String?,
      fileName: json['fileName'] as String?,
      contentType: json['contentType'] as String?,
      fileSizeBytes: json['fileSizeBytes'] as int? ?? 0,
      enteredByName: json['enteredByName'] as String? ?? '',
      createdAtUtc: DateTime.parse(json['createdAtUtc'] as String),
      downloadUrl: json['downloadUrl'] as String? ?? '',
      hasFile: json['hasFile'] as bool? ?? false,
    );
  }
}

File: ClinicNow.Services/Database/Entities/LabFinding.cs
using System.ComponentModel.DataAnnotations.Schema;

namespace ClinicNow.Services.Database.Entities;

/// <summary>
/// A lab finding ("laboratorijski nalaz") - review item C4. Unlike a generic
/// <see cref="MedicalDocument"/>, this is deliberately tied to **both** the
/// patient and the specific <see cref="Appointment"/> it was produced for (the
/// prijava's requirement the reviewer names explicitly), with a structured
/// <see cref="Result"/> field rather than being just a file with an optional
/// note. Entered by a doctor or lab staff during/after an examination;
/// legally-retained health data - soft-delete only, never a hard DELETE.
///
/// <see cref="PatientId"/> is denormalized from <see cref="Appointment.PatientId"/>
/// at creation time (derived server-side in <c>LabFindingService.CreateAsync</c>,
/// never taken from the client) rather than requiring every query to join
/// through <see cref="Appointment"/> to find a patient's lab history.
/// </summary>
public class LabFinding : ISoftDelete
{
    public int Id { get; set; }

    public int PatientId { get; set; }

    /// <summary>Optional - a soft-deleted (archived) patient's findings must stay readable, same reasoning as <c>MedicalDocument.Patient</c> (review item C3).</summary>
    public Patient? Patient { get; set; }

    public int AppointmentId { get; set; }
    public Appointment Appointment { get; set; } = null!;

    /// <summary>Which test was performed, e.g. "Kompletna krvna slika (KKS)". Required - it is what the finding *is*.</summary>
    public string TestName { get; set; } = string.Empty;

    /// <summary>The measured value, e.g. "13.9". Optional: a descriptive finding (an imaging report, say) has no single number.</summary>
    public string? Value { get; set; }

    /// <summary>Unit of measure for <see cref="Value"/>, e.g. "g/dL".</summary>
    public string? Unit { get; set; }

    /// <summary>The laboratory's normal range for this test, e.g. "12.0 - 16.0", so the value can be read in context.</summary>
    public string? ReferenceRange { get; set; }

    /// <summary>The finding/interpretation text (e.g. "Kompletna krvna slika - uredni parametri") - what the numbers mean.</summary>
    public string Result { get; set; } = string.Empty;

    /// <summary>The doctor's own remark on the finding - distinct from <see cref="Result"/>, which states the finding itself.</summary>
    public string? DoctorNote { get; set; }

    /// <summary>
    /// The attached document, if any. Optional throughout (prijava: "uz nalaz se
    /// <em>može</em> priložiti i dokument u PDF ili slikovnom formatu") - a finding
    /// entered straight from a lab report still stands on its structured fields
    /// alone. All four move together: either every one is set, or none is.
    /// </summary>
    public string? FileName { get; set; }
    public string? ContentType { get; set; }
    public byte[]? FileData { get; set; }
    public long FileSizeBytes { get; set; }

    /// <summary>True when a document is attached and the download endpoint has something to serve. Derived, never a column.</summary>
    [NotMapped]
    public bool HasFile => FileData != null && FileData.Length > 0;

    /// <summary>
    /// SHA-256 of <see cref="FileData"/>, stored at upload time and served as the
    /// download's ETag. Kept as a column rather than hashed per request: the point
    /// of the ETag is to avoid sending the bytes, so reading them back to hash them
    /// would trade the bandwidth saving for exactly the I/O it was meant to avoid.
    /// Null only for rows written before this column existed.
    /// </summary>
    public string? ContentHash { get; set; }

    public int EnteredByUserId { get; set; }
    public User EnteredByUser { get; set; } = null!;

    public DateTime CreatedAtUtc { get; set; }

    public bool IsDeleted { get; set; }
    public DateTime? DeletedAtUtc { get; set; }
}
